### ISSUE TITLE
Global CSS imported on one page gets imported on all pages

### ISSUE THREAD

**Issue by amakk** [gatsbyjs/gatsby#14333]
## Description

Hi, I am migrating and combining some older sites into one gatsby site. I need to import global CSS on some pages, but not on others. The issue is that if I import the CSS on one page, it gets included in all the other pages. 

I'm importing the CSS from the specific page that needs it, so I would expect it to show only on that page. Is this a bug?

I've seen [this issue](https://github.com/gatsbyjs/gatsby/issues/7702) but the suggested solution there is to use css modules, and migrating to CSS modules would be very time consuming at this point (initial migration phase).

I've also tried the solution [here](https://github.com/gatsbyjs/gatsby/issues/13976#issuecomment-491559103) but this inlines the CSS, whereas I'd like to have it as a separate file. 

### Steps to reproduce

1. Create a css file `styles.css`
2. Import css file into one page `import '../styles.css`
3. Open a different page and see that `styles.css` is included there too.

**Edit:** I created a repo to demonstrate the issue here: https://github.com/amakk/gatsby-global-css-bug

`index.js` imports `style.css` while `page2.js` doesn't import it. Yet, both get the style from `style.css` applied.

### Expected result

The CSS imported on one page should only be included into that page, not on other pages.

### Actual result

All pages include the CSS.

### Environment

  System:
    OS: macOS 10.14
    CPU: (8) x64 Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 8.11.3 - /usr/local/bin/node
    Yarn: 1.16.0 - /usr/local/bin/yarn
    npm: 6.4.1 - /usr/local/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 74.0.3729.169
    Firefox: 66.0.3
    Safari: 12.0
  npmPackages:
    gatsby: ^2.0.65 => 2.5.5 
    gatsby-image: ^2.0.23 => 2.1.0 
    gatsby-plugin-google-tagmanager: ^2.0.7 => 2.0.13 
    gatsby-plugin-manifest: ^2.0.11 => 2.1.1 
    gatsby-plugin-offline: ^2.0.19 => 2.1.1 
    gatsby-plugin-postcss: ^2.0.2 => 2.0.7 
    gatsby-plugin-react-helmet: ^3.0.4 => 3.0.12 
    gatsby-plugin-sass: ^2.0.11 => 2.0.11 
    gatsby-plugin-sharp: ^2.0.15 => 2.0.37 
    gatsby-plugin-styled-components: ^3.0.4 => 3.0.7 
    gatsby-plugin-webpack-bundle-analyzer: ^1.0.3 => 1.0.5 
    gatsby-source-filesystem: ^2.0.12 => 2.0.36 
    gatsby-transformer-sharp: ^2.1.9 => 2.1.19 
  npmGlobalPackages:
    gatsby-cli: 2.5.15

**Comment by LekoArts** [comment]
Hi!

Sorry to hear you're running into an issue. To help us best begin debugging the underlying cause, it is incredibly helpful if you're able to create a [minimal reproduction][reproduction]. This is a simplified example of the issue that makes it clear and obvious what the issue is and how we can begin to debug it.

If you're up for it, we'd very much appreciate if you could [provide a minimal reproduction][reproduction] and we'll be able to take another look.

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Comment by amakk** [comment]
Hi @LekoArts, I created a repository to reproduce the issue here: https://github.com/amakk/gatsby-global-css-bug. 

`index.js` imports `style.css` while `page2.js` doesn't import it. Yet, both get the style from `style.css` applied.

Thanks for your help!

**Comment by vinzun** [comment]
CSS is global by nature. You can fix it by wrapping it with a unique class (.someclass h1), the problem is... It is too long. That is why I love sass, they can do nesting.

**Comment by ImedAdel** [comment]
A possible solution is to use CSS Modules by installing [PostCSS](https://www.gatsbyjs.org/packages/gatsby-plugin-postcss/). After installing, you will have to append `.module.css` to the file name. (`index.css` => `index.module.css`) and import it as usual.

**Comment by akki-mali** [comment]
Facing the same issue for multiple layouts using scss.
I have two scss files style1.scss and style2.scss and I have imported style1.scss in layout1.js and style2.scss in layout2.js. But in the DOM for any layout both the scss files are overriding.

**Comment by amakk** [comment]
Thanks for the suggestions @vinzun and @ImedAdel. Unfortunately, these wouldn't work in my case, mainly for two reasons:

1. When importing something like bootstrap CSS, it modifies the html element, and probably others. Not sure if it's even possible to wrap all that with a class but even if it's possible, there is another issue
2. The size of the CSS will really increase, since I'm combining several sub-sites into one, so I wouldn't want to load the equivalent of 3 sites worth of CSS on each page. 

What I'm really after is an equivalent to including an external stylesheet via a link tag (which is how it's done in the current website) that I could add/remove on specific pages. I know I could just put the CSS in a static folder and create a link tag, but some of it is written in SASS and I would also lose out on cache busting.

Any other suggestions?

**Comment by vinzun** [comment]
oh I see. Never used bootsrap before, sorry I can't help you.

**Comment by amakk** [comment]
Does anyone else have other ideas? Here's what I'm doing for now:

`import bootstrap from '!raw-loader!bootstrap/dist/css/bootstrap.min.css'`

And then putting it into a style tag:
```
<Helmet>
        <style>{bootstrap}</style>
</Helmet>
```
But I'm wondering how efficient this is. If I put this in a layout component, will it get re-downloaded on each page navigation? Also, there isn't a way to tell which CSS file the styles are coming from in Chrome inspector.

**Comment by amakk** [comment]
@LekoArts could you update the status labels of this issue since I've added all needed info and reproduction

**Comment by superhawk610** [comment]
@amakk I can't say whether your solution is the best approach, but it's not downloading it on each page navigation, it will get it once as part of the initial JS bundle download and then use that version for subsequent page loads.

Using `raw-loader` copies the entire contents of the file to a string at build time and embeds it into the final JS output, so that `import` call is equivalent to

```
// component.js
import bootstrap from './bootstrap.min.css.js';

// bootstrap.min.css.js
export default `
  .class1 { /* ... */ }
  .class2 { /* ... */ }
  // ...
`;
```

**Comment by amakk** [comment]
@superhawk610 Thanks, that's good to know. Would appreciate any tips on a better approach.

**Comment by superhawk610** [comment]
For smaller scale styles you can make use of CSS Modules as suggested earlier. They'll only be loaded on pages where they're required - you can find more information on using them with Gatsby [here](https://www.gatsbyjs.org/docs/css-modules/).

For your use case, I would recommend writing a separate [layout component](https://www.gatsbyjs.org/docs/layout-components/) for each nested site that requires its own stylesheets, then inject the stylesheets using [react-helmet](https://github.com/nfl/react-helmet) and `<link rel="stylesheet" />` tags. This way, each stylesheet will only be loaded when the page layout requires it!

If you'd like to discuss further, feel free to open a thread on [Spectrum](https://spectrum.chat/gatsby-js?tab=posts) and link to this issue. Best of luck!

**Comment by HarshilShah** [comment]
@superhawk610 Question about this: Is there a way to do this with `gatsby-plugin-sass`? It seems to store the stylesheet with random names so not sure how to use that here

**Comment by kuworking** [comment]
In my case, I am abstracting code into different npm packages and each one of them had a global imported css file

In the main gatsby website, each package was used on different pages, but all global css files were downloaded/injected in all pages (also in those not using the packages at all)

I'm now deciding how to avoid using global css files in the packages, but since I've found this issue I comment this for future readers

**Comment by armandoborge** [comment]
@amakk did you manage to find a solution? I am having a similar problem trying to generate sites based on different templates but using the same repository, the CSS problem is the only sho stopper. 

Appreciate your feedback!

**Comment by armandoborge** [comment]
I've ended up with a bunch of updates to tackle the issue of having isolated templates, is not the best approach I am sure, but I'll share what I have in case it helps somebody:

1. I am using CSS modules, so I've added a main.module.css file which contains my global classes and custom properties, something like this:

```
.theme {
    --color-main: red;
    .title {
        color: var(--color-red);
    }
}
```

2. I am passing that wrapper class to react-helmet so it can be assigned to the body element

```
       import main from main.module.css

        <Helmet
            bodyAttributes={{ class:  main.theme }}
        />
```

3. And finally to avoid load the fonts of each template in all the pages, I've created separated CSS font loader files in the static folder and loading these fonts using React Helmet as well

 ```
       import main from main.module.css
       const fontsPath = 'theme-fonts-folder'

        <Helmet bodyAttributes={{ class:  main.theme }}>
                <link rel='stylesheet' type='text/css' href={`/fonts/${fontsPath}/index.css`} />
        </Helmet>
```

**Comment by uuykay** [comment]
I am also experiencing a similar issue using SCSS. I have 2 different pages. My use case is, I want page 1 to have some SCSS that the other doesn't. However SCSS imported on the page level is propagated to all pages. How is this being ignored and the issue closed?

**Comment by uuykay** [comment]
> Does anyone else have other ideas? Here's what I'm doing for now:
> 
> `import bootstrap from '!raw-loader!bootstrap/dist/css/bootstrap.min.css'`
> 
> And then putting it into a style tag:
> 
> ```
> <Helmet>
>         <style>{bootstrap}</style>
> </Helmet>
> ```
> 
> But I'm wondering how efficient this is. If I put this in a layout component, will it get re-downloaded on each page navigation? Also, there isn't a way to tell which CSS file the styles are coming from in Chrome inspector.

I did a variation of this, using a link tag, which I think would be better since the asset will be cached in the user's browser rather than having to be passed over HTTPS in the page.


1) Use a build tool to compile down to CSS (e.g. node-sass) to your static folder
```
node-sass --output-style compressed ./src/styles/style.scss ./static/styles/style-homepage.css
```

2) Use React Helmet to import the CSS in the page you are after.

```
      <Helmet>
        <link rel="stylesheet" href="/styles/style-homepage.css" />
      </Helmet>
```

**Comment by kashesandr** [comment]
Faced with similar issue. Wanted to develop admin page with another bootstrap theme but....

**Comment by shobz** [comment]
> Faced with similar issue. Wanted to develop an admin section with another bootstrap theme but....

Same issues here. Can't have two different layouts as both compiles at same. Even react helmet doesn't work with client side routes and pages.

**Comment by phun-ky** [comment]
Maybe unrelated, but I've used the "rewrite to link tag on build" approach to use external stylesheet, but I still want to use some inline styles for only the core layout, how can I approach this without manually setting style tags in the html.js file?

**Comment by ericmasiello** [comment]
Why is this issue closed? It doesn't appear like there's a real answer yet

**Comment by DuartePessoaGodinho** [comment]
I'm having the same initial issue described.
I am using sass and the idea of having to compile it down to css separately kind breaks the purpose of using a framework like Gatsby.

When importing a .scss file or .css on any file, it becomes available globally.
I don't think this is normal behaviour.

Could you re-open this Issue or at least give a final answer/work-around that actually makes sense ?

**Comment by uuykay** [comment]
Let's be real Gatsby never thought long term about a scalable asset breakdown strategy. To not have a stance on this speaks volumes. GraphQL as a data layer was a con, biggest piece of overhead at no real benefit.

**Comment by the-me-0** [comment]
Okay, I just ran into this issue. Well... after some hours of searching, as this is so dumb, excuse my words, that it had not come to my mind until very recently. I was trying to add sass modules to my style files, but as I was going I found out that the "@use" import did not import files only one time, as the doc said it should. Every file that had a " @use 'base' " would end up duplicating the style contained into my base to the css of the page I was working on. After some time searching on a fault/error on sass side, I understood it was not their side that was faulty. It's yours, gatsby.
It reminded me the other day, I was working on my header when I realized I had not imported the header style file in the page I was working, but it's style was still to be found. I looked, and told myself I must have missed it, and skipped that fact as it was not very important to me that time.

Now the fact is that saying "CSS is global by nature" is bullshit. If I want to have a dedicated css file for each of my page, I should be able to. You've taken that right away from us by combining all CSS/SCSS imported in the project into one single file, that your app serves for every single page, without taking an interest in whether this css was intended to be there.

Alright, now we could argue that it makes things easier, and that we can work around this be wrapping each of our components into a class and wrap the css for it into the same class. But this would not fix, _at least,_ those 3 issues :
- SASS modules with imported @use gets imported multiple times anyway,
- for every single page, you send out a heavy full of unused css style,
- (i forgot the third argument as I was writing, I'm really sorry)

So, until I find a great workaround or that you _at least_ re-open this issue and understand that yes, you are doing something wrong in this style files handling, I will stay on that bitter taste about gatsby.

**Comment by the-me-0** [comment]
So about this, I think I found a solution that's working for me and my SASS setup. Not that I understand how we can ever come to the fact that merging all css into a single file can be a good solution, but hey, it's not like devs are open-minded on the subject.

So I had a global style file that contained variables, and global style. Each @use I added to my files, the global style was duplicated. To fix that, I cut my global scss file into 2 : one with the vars, one with the global style. I import the global style in the gatsby-browser.ts file (that I had to create), and import the vars into my scss files just as I did before. My global style is not duplicated anymore, but I suppose the variables are loaded many times still. That has no impact, except a little on performance I suppose. I think importing a single time the global style, with `@use` in your index i.e. should work just as well because of the way gatsby "works", but anyway.

I tried the gatsby-plugin-split-css just before that, but it did not fix my problem. To be honest, it did not seem to work at all, as all my css came from the same file in the browser anyway. Only the solution above seems to work, and it's a shame it's not more accessible. From a specific issue that you would think comes from sass, it's hard to go all the way to thinking about this solution. All this pain, just because gatsby does not tell, or does not tell enough, that they decided for a certainly unvalid reason to merge all css into one.
